File: 200-number-of-islands/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const [nRows, nCols] = [grid.length, grid[0].length];
    let [ans, visited] = [0, new Set()];
    
    const inBounds = (row, col) => {
        return row >= 0 && row < nRows && col >= 0 && col < nCols;
    };
    
    const valid = (row, col) => {
        return inBounds(row, col) &&
            !visited.has(`${row}-${col}`) &&
            grid[row][col] === "1";
    };
    
    const dfs = (row, col) => {
        if (!valid(row, col)) return 0;
        
        visited.add(`${row}-${col}`);
        const fundDirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
        fundDirs.forEach(([rowDiff, colDiff]) => {
            const [newRow, newCol] = [row + rowDiff, col + colDiff];
            dfs(newRow, newCol);
        });
        return 1;
    };
    
    for (let row = 0; row < nRows; row++) {
        for (let col = 0; col < nCols; col++) {
            if (!visited.has(`${row}-${col}`)) {
                ans += dfs(row, col);
            }
        }
    }
    return ans;
};

File: 200-number-of-islands/200-number-of-islands.rb
# @param {Character[][]} grid
# @return {Integer}

def in_bounds(row, col, grid)
    return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length
end

def valid(row, col, grid, visited)
    return in_bounds(row, col, grid) && !visited.include?([row, col]) && grid[row][col] == '1'
end

def dfs(row, col, grid, visited)
    return 0 if !valid(row, col, grid, visited)
    visited.add([row, col])
    dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    dirs.each do |rowDiff, colDiff|
        dfs(row+rowDiff, col+colDiff, grid, visited)
    end
    return 1
end

def num_islands(grid)
    return 0 if grid.length == 0
    visited = Set.new
    ans = 0
    (0...grid.length).each do |row|
        (0...grid[row].length).each do |col|
            if !visited.include?([row, col])
                ans += dfs(row, col, grid, visited)
            end
        end
    end
    return ans
end